### PR TITLE
Overhaul search filters

### PR DESCRIPTION
**Pull request by SamantazFox**
Replaces #2544

### CODE REVIEW

**Comment by SamantazFox** [comment]
Note to self: change translation keys in other languages

**Comment by SamantazFox** [comment]
Poke @matthewmcgarvey: do you mind reviewing that?

File: src/invidious/search/filters.cr [hunk 1]
```diff
@@ -228,6 +228,25 @@ module Invidious::Search
...
+      raw_params["date"] = [@date.to_s.underscore] if !@date.none?
+      raw_params["type"] = [@type.to_s.underscore] if !@type.all?
+      raw_params["sort"] = [@sort.to_s.underscore] if !@sort.relevance?
+
+      if !@duration.none?
+        raw_params["duration"] = [@duration.to_s.underscore]
+      end
+
+      if !@features.none?
+        raw_params["features"] = [Filters.format_features(@features)]
```
Comment (matthewmcgarvey) [inline src/invidious/search/filters.cr:249]: Should this split on the comma? It's going to look like `["live,4k,3d"]`

**Review by matthewmcgarvey — APPROVED** [review]
Looks good. Little difficult to look at everything in great detail but I'm confident this will work 👍

**Comment by SamantazFox** [comment]
> Little difficult to look at everything in great detail but I'm confident this will work 👍

If that helps, each commit is made so that it can be reviewed separately (but still in order). Checking out at any commit should compile without an error message and the CI should always pass.

File: src/invidious/search/filters.cr [hunk 2]
```diff
@@ -192,6 +192,42 @@ module Invidious::Search
       return {filters, channel, cleaned_query, subscriptions}
     end
 
+    def self.from_iv_params(params : HTTP::Params) : Filters
+      # Temporary variables
+      filters = Filters.new
+
+      if type = params["type"]?
+        filters.type = Type.parse?(type) || Type::All
+        params.delete("type")
```
[the remark this answers is not in the document]
Comment (SamantazFox) [reply src/invidious/search/filters.cr:206]: My idea is to use this as a sanitization function, by calling it from kemal's middleware (a dedicated handler, as it's done for session management). By removing the parameters, that will ensure that the non-sanitized inputs won't be used somewhere else. The goal would to end up an empty `HTTP::Params` object.

File: src/invidious/search/filters.cr (hunk 1 above)
Comment (SamantazFox) [reply src/invidious/search/filters.cr:249]: No, it shoudln't split on the comma, and yup, that's expected :)

Previously, the features filter was passed in the query, as so:
```some search query features:live,4k,3d"```

Imo it's better to keep it that way, even with URL parameters, so the URL can be easily edited by hand, without needing the lenghy `&features=live&features=4k&features=3d....`.

**Comment by SamantazFox** [comment]
I've fixed a few things since Matthew's review:
- Broken previous/next URLs on search page (forgot surrounding `<%= %>`)
- Broken `Query.process` (`when` should contain all conditions on a single line, separated by commas)
- Enhanced accessibility on filters (by using `<fieldset>` and `<legend>` HTML elements)

**Comment by SamantazFox** [comment]
Stuff that remains "TODO": properly handle incompatible search filters (e.g `duration=short` + `type=channel`). Maybe some JS to make that behavior more explicit on the client side could be nice.

**Review by matthewmcgarvey — APPROVED** [review]

**Comment by SamantazFox** [comment]
Deployed on test.invidious.io.
Bogus search query (to see new error message): https://test.invidious.io/search?q=LTT&type=channel&duration=short

**Review by matthewmcgarvey — APPROVED** [review]

**Comment by SamantazFox** [comment]
Sorry for the multiple force-pushes :x

**Pull request merged**